command_line/fp3.py

```diff
@@ -0,0 +1,578 @@
+        logger.info(scope.fetch_diff(self._working).as_str())
+
+        self._crystal = None
+        self._root = os.getcwd()
+        self._n = nproc()
+
+        # quick checks...
+        scan = self._experiment[0].scan
+        osc = scan.get_oscillation()
+        rng = scan.get_image_range()
+
+        self._osc = osc
+        self._rng = rng
+
+        self._integrated = []
+        self._combined = None
+        self._symmetry = None
+        self._scaled = None
+
+    def _write_phil(self, what, where):
+        """helper function: extract (what) from global scope and write as
+        (where)/(what).phil iff it contains anything. If file written,
+        returns [filename] else [] to allow straight command-line adding."""
+
+        phil = format_phil_include(scope, self._working, what).strip()
+        if not phil:
+            return []
+
+        out = os.path.join(where, f"{what}.phil")
+        with open(out, "w") as f:
+            f.write(phil)
+        return [out]
+
+    def index(self):
+        """Initial find spots and indexing: if it has been run already will
+        just reload the results from the previous run."""
+
+        work = os.path.join(self._root, "index")
+        if os.path.exists(work):
+            indexed = ExperimentList.from_file(os.path.join(work, "indexed.expt"))
+
+            self._experiment[0].crystal = indexed[0].crystal
+            logger.info("Picked up pre-exising crystal:")
+            logger.info(indexed[0].crystal)
+            return
+
+        if not os.path.exists(work):
+            os.mkdir(work)
+
+        self._experiment.as_file(os.path.join(work, "input.expt"))
+
+        five = int(round(5 / self._osc[1]))
+        i0, i1 = self._rng
+        blocks = [(b[0] + 1, b[1]) for b in index_blocks(i0 - 1, i1, self._osc[1])]
+
+        logger.info("Finding spots...")
+        phil = self._write_phil("find_spots", work)
+        result = procrunner.run(
+            ["dials.find_spots", "input.expt", "nproc=%d" % self._n]
+            + phil
+            + ["scan_range=%d,%d" % block for block in blocks],
+            working_directory=work,
+            print_stdout=self._debug,
+            print_stderr=self._debug,
+        )
+
+        logger.info("Indexing...")
+        # let's just assume that was fine - so index
+
+        phil = self._write_phil("index", work)
+        result = procrunner.run(
+            ["dials.index", "input.expt", "strong.refl"] + phil,
+            working_directory=work,
+            print_stdout=self._debug,
+            print_stderr=self._debug,
+        )
+
+        indexed = ExperimentList.from_file(os.path.join(work, "indexed.expt"))
+        logger.info(indexed[0].crystal)
+        self._experiment[0].crystal = indexed[0].crystal
+
+    def integrate(self):
+        """Integration of the complete scan: will split the data into 5 deg
+        chunks and spin the integration of each chunk off separately"""
+
+        rng = self._rng
```

Comment:
the naming here is a bit terse. I was thinking "random number generator", not "image range"